src/serial/MidiSessionALSA.cc

```diff
@@ -166,13 +174,253 @@ void MidiOutALSA::serialize(Archive& /*ar*/, unsigned /*version*/)
+	, name(snd_seq_client_info_get_name(&cinfo))
+	, desc(snd_seq_port_info_get_name(&pinfo))
+{
+	eventDistributor.registerEventListener(EventType::MIDI_IN_ALSA, *this);
+	if ((snd_seq_port_info_get_capability(&pinfo) & (SND_SEQ_PORT_CAP_WRITE | SND_SEQ_PORT_CAP_SUBS_WRITE))) {
+		name.append(" (input)");
+	}
+}
+
+MidiInALSA::~MidiInALSA()
+{
+	if (connected) {
+		disconnect();
+	}
+	eventDistributor.unregisterEventListener(EventType::MIDI_IN_ALSA, *this);
+}
+
+void MidiInALSA::plugHelper(Connector& connector_, EmuTime::param /*time*/)
+{
+	auto& midiConnector = checked_cast<MidiInConnector&>(connector_);
+	midiConnector.setDataBits(SerialDataInterface::DATA_8); // 8 data bits
+	midiConnector.setStopBits(SerialDataInterface::STOP_1); // 1 stop bit
+	midiConnector.setParityBit(false, SerialDataInterface::EVEN); // no parity
+
+	setConnector(&midiConnector); // base class will do this in a moment,
+	                              // but thread already needs it
+	connect();
+}
+
+void MidiInALSA::unplugHelper(EmuTime::param /*time*/)
+{
+	if (connected) {
+		disconnect();
+	}
+}
+
+void MidiInALSA::connect()
+{
+	destinationPort = snd_seq_create_simple_port(
+		&seq, "MIDI in pluggable",
+		SND_SEQ_PORT_CAP_WRITE | SND_SEQ_PORT_CAP_SUBS_WRITE, SND_SEQ_PORT_TYPE_MIDI_GENERIC);
+	if (destinationPort < 0) {
+		throw PlugException(
+			"Failed to create ALSA port: ", snd_strerror(destinationPort));
+	}
+
+	int err = snd_seq_connect_from(&seq, destinationPort, srcClient, srcPort);
+	if (err) {
+		snd_seq_delete_simple_port(&seq, destinationPort);
+		throw PlugException(
+			"Failed to connect to ALSA port "
+			"(", srcClient, ':', srcPort, ")"
+			": ", snd_strerror(err));
+	}
+
+	snd_midi_event_new(MidiOutDevice::MAX_MESSAGE_SIZE, &event_parser);
+	snd_midi_event_no_status(event_parser, 1);
+
+	connected = true;
+	stop = false;
+
+	thread = std::thread([this]() { run(); });
+}
+
+void MidiInALSA::disconnect()
+{
+	stop = true;
+	thread.join();
+
+	snd_midi_event_free(event_parser);
+	snd_seq_disconnect_from(&seq, destinationPort, srcClient, srcPort);
+	snd_seq_delete_simple_port(&seq, destinationPort);
+
+	connected = false;
+}
+
+void MidiInALSA::run()
+{
+	assert(isPluggedIn());
+
+	auto npfd = snd_seq_poll_descriptors_count(&seq, POLLIN);
+	std::vector<struct pollfd> pfd(npfd);
+	snd_seq_poll_descriptors(&seq, pfd.data(), npfd, POLLIN);
+
+	while (true) {
+		if (poll(pfd.data(), npfd, 1000) > 0) {
+			do {
+				snd_seq_event_t *ev = NULL;
+				long err, size;
+
+				if ((err = snd_seq_event_input(&seq, &ev)) < 0) {
+					std::cerr << "Error receiving MIDI event: "
+						<< snd_strerror(err) << '\n';
+					break;
+				}
+
+				if (!ev) {
+					break;
+				}
+
+				if (ev->type == SND_SEQ_EVENT_SYSEX) {
+					std::lock_guard<std::mutex> lock(mutex);
+
+					for (auto i : xrange(ev->data.ext.len)) {
+						queue.push_back(static_cast<uint8_t*>(ev->data.ext.ptr)[i]);
+					}
+				} else {
+					uint8_t bytes[12] = {};
```

Comment:
Very much a detail / coding style:
Recently (last couple of weeks) I replaced all C-style arrays with c++ std::array. Rationale: c++ inherited arrays from the C language, including to associated problems (implicit array-to-pointer decay,  can't pass/return an array by value to/from a function (instead you pass a pointer to the array), ...). std::array solves some of these problems.

This line then becomes: `std::array<uint8_t, 12> bytes = {};`
And because there is no implicit array-to-pointer decay anymore, the call below then becomes:
`snd_midi_event_decode(event_parser, bytes.data(), bytes.size(), ev);`
The usage further down `queue.push_back(bytes[i]);` remains unchanged.

Reply:
Done.